### ISSUE TITLE
bindfilter, i.e. expand userattr to a full LDAP filter

### ISSUE THREAD

**Issue by ghenry** [hashicorp/vault#5304]
**Is your feature request related to a problem? Please describe.**

At SureVoIP we do an initial LDAP search with an ldap entity that only has limited permissions to search to find a valid user DN and then re-bind with the supplied password. Your LDAP config almost allows this. Instead of just passing (userattr=uid) as a filter, it would be ideal for that to be supplied as a full LDAP filter, like (uid=%s) etc. where you replace with sprintf. For example from one of our Apps and used in all our services (user_filter):

          <store>
                class LDAP
                ldap_server xxx
                <ldap_server_options>
                        timeout 30
                </ldap_server_options>
                binddn xxx
                bindpw xxx
                start_tls 1
                user_basedn xxx
                user_filter (&(objectClass=person)(uid=%s)(SuretecAccountActive=TRUE)(SuretecServiceActive=TRUE)(SuretecServiceEnabled=SUREVOIPPORTAL))
                user_scope one
                user_field xxx
                <user_search_options>
                    deref always
                </user_search_options>
            </store>

This means we can toggle any things in the filter to FALSE and kick out a user or let them into a service.
 
**Describe the solution you'd like**
rename userattr to userfilter in config and update code here:

https://github.com/hashicorp/vault/blob/b65961e3b695105192b1b518b159d6c59eef0221/helper/ldaputil/config.go#L70
https://github.com/hashicorp/vault/blob/b65961e3b695105192b1b518b159d6c59eef0221/helper/ldaputil/client.go#L116

It's almost there, you just limit the filter to (userattr=username)

**Describe alternatives you've considered**
Using groups, but I don't want to let the user get that far.

**Explain any additional use-cases**
A complete LDAP solution like is possible with almost all LDAP auth options in applications.

**Additional context**
None. Happy to do a patch, test and doc update.

Thanks!

**Comment by Lasering** [comment]
I would also find this change very helpful. The `userattr` should be changed to `userfilter` with the default value of `(cn=%s)` and this line to:

`filter := fmt.Sprintf("(%s=%s)", cfg.UserAttr, ldap.EscapeFilter(username))`

Changed to:

`filter := fmt.Sprintf(cfg.UserFilter, ldap.EscapeFilter(username))`

This way the LDAP authentication would be way more powerfull. I can make a PR if it helps.

**Comment by Lasering** [comment]
Can I help making progress in this issue?

**Comment by mnsh** [comment]
Having a `userfilter` conifg option to control the search step in username-to-bind-dn-mapping would indeed be very useful.

For example, given `userattr=uid` and `userfilter=(&(objectClass=customUser)(|(uid={{.Username}})(customPrincipalName={{.Username}})))` would:

1. Limit the search to a specific objectClass instead of just going with `(uid=...)`
2. Allow logins using both `user.name` and `user.name@domain`

When `userfilter` is set, the value of `userattr` should still decide what Vault would use as the username for things like identity mapping.

**Comment by Lasering** [comment]
Would a PR help? I can do everything except the UI part.

**Comment by karras** [comment]
We have multiple projects with the same requirement. Currently it seems not possible to enforce that a user must be member of a group to login. For example: Only users of the group `G-Vault-Users` are allowed to login.

To my knowledge this is usually done through `userfilter` as mentioned above.

**Comment by ixe013** [comment]
I rebased @LiviusNL code and added documentation to it. Will test it tomorow and get the pull request started the code works. It LGTM : clean, simple, backward compatible.

**Comment by zerkms** [comment]
@ixe013 will it eventually become a PR then?

**Comment by ixe013** [comment]
@karras you are right, there are currently (as of Vault 1.6.2) no way to prevent users that are not a member of group to log in Vault. While you wait for this update to go through, you can at least make sure they are not assigned any roles, so they only see the cubbyhole mount when they do log in.

**Comment by ixe013** [comment]
> @ixe013 will it eventually become a PR then?

Yes, it will. The code looks good and I need the feature. Just can't test it tonight. +We are a Vault Enterprise customer. I'll ask our technical account manager if he can help get the PR some visibility when it is submitted.

**Comment by ixe013** [comment]
I tested the code, works as expected. I'll add the UI code and submit the PR tonight. 

Last call @LiviusNL ... I you can't sign the CLA agreement I'll push your great code under my name/responsability.

**Comment by LiviusNL** [comment]
@ixe013 Although the feature is code complete, the implementation is not complete. The tests for testing the additional functionality have not been written, which I see as a must for the feature to be incorporated. Furthermore, I did not add updated documentation.

**Comment by ixe013** [comment]
Thanks! I've added the documentation and updated the UI. I'll write some tests, too, as soon as I get Docker running on WSL2

**Comment by igorcoding** [comment]
Hello! Are there any news on this feature? Really look forward to having this implemented.

**Comment by ixe013** [comment]
Nothing moved since I created the pull request. We are a paying Vault Enterprise user, so I created a support case for someone to look at the PR. All I got is this:
> Thanks for submitting your pull request. I have forwarded it to the product manager as reference number xxxxxxxxxxxxxxx and it will be considered for future inclusion in the product.

I keep rebasing [my pull request](https://github.com/hashicorp/vault/pull/11000), like I did for the 1.7.x releases, so that a merge is "_just_" a click away for Hashicorp.

A comment on the PR and to your Vault Technical Account Manager can only help give it some visibility.
